Add text/template functions for manipulating "output" values

jippi opened the pull request:
Description
-------------

I thought adding some `text/template` functions to the "outputter" would make for a more flexible and more generic solution to #509 and other "I want my Mocks in an oddly specific path" situations where some light-weight string manipulation is necessary 

Adding `replace` (aka `strings.Replace()` and  `trimPrefix` (aka `strings.TrimPrefix`) template functions to the outputter so that it's easier to manipulate it more freely. 

My personal use case is generating mocks for `internal` packages in a somewhat structured way. #509 doesn't seem to work for me, and ideally, I would want to put manage the path in a more flexible way than initially provided by #509 - so I thought a couple of template functions could support this better without having to implement many different hacks.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X]  New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [X] This change requires a documentation update

Version of Golang used when building/testing:
---------------------------------------------

- [ ] 1.11
- [ ] 1.12
- [ ] 1.13
- [ ] 1.14
- [ ] 1.15
- [ ] 1.16
- [ ] 1.17
- [ ] 1.18
- [ ] 1.19
- [X] 1.20

How Has This Been Tested?
---------------------------

I wanted to put internal mocks in the internal folder but without the additional `internal` path segment.

```yaml
packages:
  ./internal:
    config:
      all: true
      recursive: true
      disable-version-string: true
      dir: 'internal/mocks/{{ trimPrefix .InterfaceDirRelative "internal" }}'
      filename: "{{ .InterfaceName }}.go"
      mockname: "{{ .Mock }}{{ .InterfaceName }}"
      outpkg: "{{ .PackageName }}_mocks"
```

yielding

```
$ find internal/mocks 
internal/mocks
internal/mocks/app
internal/mocks/app/letsgo
internal/mocks/app/letsgo/subcommands
internal/mocks/app/letsgo/subcommands/exec
internal/mocks/app/letsgo/subcommands/exec/TaskRunner.go
```

(without the change, it would have been `internal/mocks/internal/app/letsgo/subcommands/exec/TaskRunner.go`)

It also happens to make #509 much nicer and user-controllable via `replace`

```yaml
packages:
  ./internal:
    config:
      all: true
      recursive: true
      disable-version-string: true
      dir: 'mocks/{{ replace .InterfaceDirRelative "internal/" "internal_/" }}'
      filename: "{{ .InterfaceName }}.go"
      mockname: "{{ .Mock }}{{ .InterfaceName }}"
      outpkg: "{{ .PackageName }}_mocks"
```

yielding 

```
$ find mocks      
mocks
mocks/internal_
mocks/internal_/app
mocks/internal_/app/letsgo
mocks/internal_/app/letsgo/subcommands
mocks/internal_/app/letsgo/subcommands/exec
mocks/internal_/app/letsgo/subcommands/exec/TaskRunner.go
```

Checklist
-----------

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

LandonTClipp commented:
Thanks for the PR. I generally like this approach and I think the final solution will look a lot like this. I have a couple minor changes but I’ll elaborate later. I’m currently driving a camper trailer from Florida to Illinois so I’m preoccupied 😝

LandonTClipp commented:
Thanks for the PR. I generally like this approach and I think the final solution will look a lot like this. I have a couple minor changes but I’ll elaborate later. I’m currently driving a camper trailer from Florida to Illinois so I’m preoccupied 😝

jippi commented:
@LandonTClipp cool! :) Don't GitHub and drive 😆

LandonTClipp commented:
I like the idea of sprig but I'm a bit overwhelmed with the number of functions provided... not sure if 99% of those would realistically ever be used by users of mockery. I'm more of the opinion that we should add functions as needed like in this PR.

LandonTClipp commented on pkg/outputter.go:
```diff
@@ -197,6 +197,13 @@ func parseConfigTemplates(ctx context.Context, c *config.Config, iface *Interfac
 
+	// Functions available in the template for manipulating
+	// the individual key/value pairs from [templateMap] above
+	templateFuncMap := template.FuncMap{
+		"replace":    templateReplaceFn,
+		"trimPrefix": templateTrimPrefixFn,
+	}
```
Why not reference `strings.ReplaceAll` directly etc? Let's also call the function name `replaceAll`.

jippi replied:
Most did it like this because

1. I hacked it up at 3 am, and I was tired 😆 
2. See (1)

LandonTClipp reviewed (commented):
I think we can go ahead with adding this in, I just have a question on why you are adding a small wrapper around the `strings` functions.

jippi commented:
> I like the idea of sprig but I'm a bit overwhelmed with the number of functions provided... not sure if 99% of those would realistically ever be used by users of mockery. I'm more of the opinion that we should add functions as needed like in this PR.

We can pick and choose what functions we want, it's thankfully not an all-in-one :)

I think perhaps some string-manipulating ones would be nice as well, with a text/template you can do a lot of really interesting stuff all of the sudden

LandonTClipp commented:
> We can pick and choose what functions we want, it's thankfully not an all-in-one :)

I see, that makes it much better. Let's go ahead with the small change in this PR (including changes) and if there is demand for a lot of things that sprig solves, we might be able to integrate that.

jippi commented:
@LandonTClipp I pushed up a proposed change. I asked for some internal feedback from the Go crew at SeatGeek and took a leap and added the functions they provided as likely usable to them in the future. 

If they are not acceptable to you, I will strip them out again without any complaints. 

I added a `CONTRIBUTING.md` file to cover the steps I had to manually discover for getting things working locally with Taskfile and expanded Taskfile with `mkdocs` specific commands that I also had to discover manually. 

I hope that is okay with you; if not, I'll also strip it out.

It looks like the gopls formatter made additional changes in unrelated code, like trimming spaces and using octets; those seem good to me, but again, I will remove them if you don't want them.

LandonTClipp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+## Local development setup
+
+This project uses [Taskfile](https://taskfile.dev/), please see the [Install Taskfile](https://taskfile.dev/installation/) and
+[Taskfile usage documentation](https://taskfile.dev/usage/) for more information.
+
+Once installed, run `task -l` for list of valid targets.
```
You shouldn't have to reference the installation docs for taskfile, we actually import all tools mockery uses in the `tools/` sub-module so you can just do `go mod download -x` to get everything.

LandonTClipp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+## Working with documentation
+
+We use [mkdocs](https://www.mkdocs.org/) with the [mkdocs-material theme](https://squidfunk.github.io/mkdocs-material/).
+
+To preview the documentation locally, run `task mkdocs.serve`. The task will install the required mkdocs plugins and theme
+and run the mkdocs server with real-time updating/refreshing.
```
This is good 👍🏻

jippi replied:
Thank you 🙏

LandonTClipp commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+## Local development setup
+
+This project uses [Taskfile](https://taskfile.dev/), please see the [Install Taskfile](https://taskfile.dev/installation/) and
```
You shouldn't have to reference the installation docs for taskfile, we actually import all tools mockery uses in the tools/ sub-module so you can just do `go mod download -x` to get everything.

LandonTClipp replied:
It just downloads all the dependencies listed in go.mod. The tools submodule contains the tools in its go.mod file so asking go to download the deps should take care of it for you. 

No you can keep it, it's still useful.

LandonTClipp commented on docs/configuration.md:
```diff
@@ -108,3 +108,37 @@ Please see the [features section](/mockery/features/#packages-configuration) for
+
+To learn more about the templating syntax, please [see the Go `text/template` documentation](https://pkg.go.dev/text/template){target="_blank"}
+
+* [`contains` string substr](https://pkg.go.dev/strings#Contains){target="_blank"}
```
Hmm, maybe we should find a way to have mkdocs open all links in new tabs. That way we wouldn't have to do the `target="_blank"`. This is fine for now, it's a PR for another day.

jippi commented on Taskfile.yml:
```diff
@@ -22,26 +22,37 @@ tasks:
+  mkdocs.install:
+    desc: install mkdocs and plugins
+    cmds:
+      - pip install mkdocs mkdocs-glightbox "mkdocs-material>=9.1.15" pillow cairosvg
```
God dammit, can't believe I missed that lol

jippi commented on docs/configuration.md:
```diff
@@ -108,3 +108,37 @@ Please see the [features section](/mockery/features/#packages-configuration) for
+!!! note
+    Templated functions are only available when using the `packages` config feature.
+
+Template functions allow you to inspect and manipulate template variables.
```
Fixed 👌

jippi commented:
I _think_ all feedback has been actioned 💪

LandonTClipp commented:
Looks good, if you could just squash down to 1 commit with a good description, I'll approve and merge. I'll ignore the coverage test that's failing, I don't think we need tests for this.

jippi commented:
There you go @LandonTClipp :)

Busterkinng88 commented:
🤴
